Panic in govc dvs.portgroup.info on version 0.16.0 and 0.17.0 but not seen in 0.15.0

Command still actually returns expected results but panics at the end:
0.16.0
```
panic: interface conversion: types.BaseVmwareDistributedVirtualSwitchVlanSpec is *types.VmwareDistributedVirtualSwitchTrunkVlanSpec, not *types.VmwareDistributedVirtualSwitchVlanIdSpec

goroutine 1 [running]:
github.com/vmware/govmomi/govc/dvs/portgroup.(*info).Run(0xc420077110, 0x18ad580, 0xc42007e010, 0xc4200944e0, 0x0, 0x0)
	/home/vmware/release/src/github.com/vmware/govmomi/govc/dvs/portgroup/info.go:144 +0xc16
github.com/vmware/govmomi/govc/cli.Run(0xc420088050, 0x3, 0x3, 0xc88700)
	/home/vmware/release/src/github.com/vmware/govmomi/govc/cli/command.go:156 +0x522
main.main()
	/home/vmware/release/src/github.com/vmware/govmomi/govc/main.go:82 +0x63
```

0.17.0
```
panic: interface conversion: types.BaseVmwareDistributedVirtualSwitchVlanSpec is *types.VmwareDistributedVirtualSwitchTrunkVlanSpec, not *types.VmwareDistributedVirtualSwitchVlanIdSpec

goroutine 1 [running]:
github.com/vmware/govmomi/govc/dvs/portgroup.(*info).Run(0xc4200a3200, 0x110f8c0, 0xc420036058, 0xc4200805a0, 0x0, 0x0)
	/home/vmware/release/src/github.com/vmware/govmomi/govc/dvs/portgroup/info.go:144 +0xb71
github.com/vmware/govmomi/govc/cli.Run(0xc4200300d0, 0x3, 0x3, 0xc420098058)
	/home/vmware/release/src/github.com/vmware/govmomi/govc/cli/command.go:156 +0x53b
main.main()
	/home/vmware/release/src/github.com/vmware/govmomi/govc/main.go:85 +0x63
```

--- a/govc/dvs/portgroup/info.go
+++ b/govc/dvs/portgroup/info.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"flag"
 	"fmt"
+	"io"
 
 	"github.com/vmware/govmomi/govc/cli"
 	"github.com/vmware/govmomi/govc/flags"
@@ -37,7 +38,7 @@ type info struct {
 	inside     bool
 	uplinkPort bool
 	vlanID     int
-	count      uint
+	count      int
 }
 
 func init() {
@@ -54,15 +55,7 @@ func (cmd *info) Register(ctx context.Context, f *flag.FlagSet) {
 	f.BoolVar(&cmd.inside, "inside", true, "Filter by port inside or outside status")
 	f.BoolVar(&cmd.uplinkPort, "uplinkPort", false, "Filter for uplink ports")
 	f.IntVar(&cmd.vlanID, "vlan", 0, "Filter by VLAN ID (0 = unfiltered)")
-	f.UintVar(&cmd.count, "count", 0, "Number of matches to return (0 = unlimited)")
-}
-
-func (cmd *info) Process(ctx context.Context) error {
-	if err := cmd.ClientFlag.Process(ctx); err != nil {
-		return err
-	}
-
-	return nil
+	f.IntVar(&cmd.count, "count", 0, "Number of matches to return (0 = unlimited)")
 }
 
 func (cmd *info) Usage() string {
@@ -74,7 +67,37 @@ func (cmd *info) Description() string {
 
 Examples:
   govc dvs.portgroup.info DSwitch
+  govc dvs.portgroup.info -pg InternalNetwork DSwitch
   govc find / -type DistributedVirtualSwitch | xargs -n1 govc dvs.portgroup.info`
+}
+
+type infoResult struct {
+	Port []types.DistributedVirtualPort
+	cmd  *info
+}
+
+func (r *infoResult) Write(w io.Writer) error {
+	for _, port := range r.Port {
+		var vlanID int32
+		setting := port.Config.Setting.(*types.VMwareDVSPortSetting)
+
+		switch vlan := setting.Vlan.(type) {
+		case *types.VmwareDistributedVirtualSwitchVlanIdSpec:
+			vlanID = vlan.VlanId
+		case *types.VmwareDistributedVirtualSwitchTrunkVlanSpec:
+		case *types.VmwareDistributedVirtualSwitchPvlanSpec:
+			vlanID = vlan.PvlanId
+		}
+
+		// Show port info if: VLAN ID is not defined, or VLAN ID matches requested VLAN
+		if r.cmd.vlanID == 0 || vlanID == int32(r.cmd.vlanID) {
+			fmt.Printf("PortgroupKey: %s\n", port.PortgroupKey)
+			fmt.Printf("DvsUuid:      %s\n", port.DvsUuid)
+			fmt.Printf("VlanId:       %d\n", vlanID)
+			fmt.Printf("PortKey:      %s\n\n", port.Key)
+		}
+	}
+	return nil
 }
 
 func (cmd *info) Run(ctx context.Context, f *flag.FlagSet) error {
@@ -131,34 +154,20 @@ func (cmd *info) Run(ctx context.Context, f *flag.FlagSet) error {
 		criteria.PortgroupKey = []string{dvp.Key}
 	}
 
-	res, err := dvs.FetchDVPorts(ctx)
+	res, err := dvs.FetchDVPorts(ctx, &criteria)
 	if err != nil {
 		return err
 	}
 
-	var returnedPorts uint
-
-	// Iterate over returned ports
-	for _, port := range res {
-		portConfigSetting := port.Config.Setting.(*types.VMwareDVSPortSetting)
-		portVlan := portConfigSetting.Vlan.(*types.VmwareDistributedVirtualSwitchVlanIdSpec)
-		portVlanID := portVlan.VlanId
-
-		// Show port info if: VLAN ID is not defined, or VLAN ID matches requested VLAN
-		if cmd.vlanID == 0 || portVlanID == int32(cmd.vlanID) {
-			returnedPorts++
-
-			fmt.Printf("PortgroupKey: %s\n", port.PortgroupKey)
-			fmt.Printf("DvsUuid:      %s\n", port.DvsUuid)
-			fmt.Printf("VlanId:       %d\n", portVlanID)
-			fmt.Printf("PortKey:      %s\n\n", port.Key)
-
-			// If we are limiting the count and have reached the count, then stop returning output
-			if cmd.count > 0 && returnedPorts == cmd.count {
-				break
-			}
-		}
+	// Truncate output to -count if specified
+	if cmd.count > 0 && cmd.count < len(res) {
+		res = res[:cmd.count]
 	}
 
-	return nil
+	info := infoResult{
+		cmd:  cmd,
+		Port: res,
+	}
+
+	return cmd.WriteResult(&info)
 }

--- a/govc/flags/version.go
+++ b/govc/flags/version.go
@@ -21,7 +21,7 @@ import (
 	"strings"
 )
 
-const Version = "0.17.0"
+const Version = "0.17.1"
 
 type version []int
 

--- a/govc/test/network.bats
+++ b/govc/test/network.bats
@@ -176,6 +176,12 @@ load test_helper
   info=$(govc dvs.portgroup.info "$id" | grep VlanId: | uniq | grep 3123)
   [ -n "$info" ]
 
+  info=$(govc dvs.portgroup.info -json "$id" | jq  '.Port[].Config.Setting.Vlan | select(.VlanId == 3123)')
+  [ -n "$info" ]
+
+  info=$(govc dvs.portgroup.info -json "$id" | jq  '.Port[].Config.Setting.Vlan | select(.VlanId == 7777)')
+  [ -z "$info" ]
+
   run govc object.destroy "network/${id}-ExternalNetwork" "network/${id}-InternalNetwork" "network/${id}"
   assert_success
 }

--- a/govc/test/object.bats
+++ b/govc/test/object.bats
@@ -158,7 +158,8 @@ load test_helper
   assert_matches dvportgroup-
 
   run govc object.collect -s -type DistributedVirtualPortgroup / config.name
-  assert_success DC0_DVPG0
+  assert_matches DC0_DVPG0
+  assert_matches DVS0-DVUplinks-
 
   run govc object.collect -s -type DistributedVirtualPortgroup / effectiveRole
   assert_number
@@ -292,7 +293,7 @@ load test_helper
   assert_number
 
   run govc object.collect -s -type Network / summary.accessible
-  assert_success "$(printf "true\ntrue")"
+  assert_success "$(printf "true\ntrue\ntrue")"
 
   run govc object.collect -s -type Network / summary.ipPoolName
   assert_success ""

--- a/object/distributed_virtual_switch.go
+++ b/object/distributed_virtual_switch.go
@@ -66,9 +66,10 @@ func (s DistributedVirtualSwitch) AddPortgroup(ctx context.Context, spec []types
 	return NewTask(s.Client(), res.Returnval), nil
 }
 
-func (s DistributedVirtualSwitch) FetchDVPorts(ctx context.Context) ([]types.DistributedVirtualPort, error) {
+func (s DistributedVirtualSwitch) FetchDVPorts(ctx context.Context, criteria *types.DistributedVirtualSwitchPortCriteria) ([]types.DistributedVirtualPort, error) {
 	req := &types.FetchDVPorts{
-		This: s.Reference(),
+		This:     s.Reference(),
+		Criteria: criteria,
 	}
 
 	res, err := methods.FetchDVPorts(ctx, s.Client(), req)

--- a/simulator/dvs.go
+++ b/simulator/dvs.go
@@ -63,6 +63,12 @@ func (s *DistributedVirtualSwitch) AddDVPortgroupTask(c *types.AddDVPortgroup_Ta
 				VmVnicNetworkResourcePoolKey: spec.VmVnicNetworkResourcePoolKey,
 			}
 
+			if pg.Config.DefaultPortConfig == nil {
+				pg.Config.DefaultPortConfig = &types.VMwareDVSPortSetting{
+					Vlan: new(types.VmwareDistributedVirtualSwitchVlanIdSpec),
+				}
+			}
+
 			pg.PortKeys = []string{}
 
 			s.Portgroup = append(s.Portgroup, pg.Self)
@@ -113,10 +119,13 @@ func (s *DistributedVirtualSwitch) ReconfigureDvsTask(req *types.ReconfigureDvs_
 					Map.AddReference(pg, &pg.Host, member.Host)
 				}
 			case types.ConfigSpecOperationRemove:
-				if pg := FindReference(host.Network, s.Portgroup...); pg != nil {
-					return nil, &types.ResourceInUse{
-						Type: pg.Type,
-						Name: pg.Value,
+				for _, ref := range host.Vm {
+					vm := Map.Get(ref).(*VirtualMachine)
+					if pg := FindReference(vm.Network, s.Portgroup...); pg != nil {
+						return nil, &types.ResourceInUse{
+							Type: pg.Type,
+							Name: pg.Value,
+						}
 					}
 				}
 
@@ -171,10 +180,6 @@ func (s *DistributedVirtualSwitch) dvPortgroups(_ *types.DistributedVirtualSwitc
 				Setting: pg.Config.DefaultPortConfig,
 			},
 		})
-
-		if pg.PortKeys == nil {
-			continue
-		}
 
 		for _, key := range pg.PortKeys {
 			res = append(res, types.DistributedVirtualPort{

--- a/simulator/dvs_test.go
+++ b/simulator/dvs_test.go
@@ -29,7 +29,6 @@ import (
 
 func TestDVS(t *testing.T) {
 	m := VPX()
-	m.Portgroup = 0 // disabled DVS creation
 
 	defer m.Remove()
 
@@ -46,10 +45,11 @@ func TestDVS(t *testing.T) {
 	finder.SetDatacenter(dc[0])
 	folders, _ := dc[0].Folders(ctx)
 	hosts, _ := finder.HostSystemList(ctx, "*/*")
+	dvs0 := object.NewDistributedVirtualSwitch(c, Map.Any("DistributedVirtualSwitch").Reference())
 
 	var spec types.DVSCreateSpec
 	spec.ConfigSpec = &types.VMwareDVSConfigSpec{}
-	spec.ConfigSpec.GetDVSConfigSpec().Name = "DVS0"
+	spec.ConfigSpec.GetDVSConfigSpec().Name = "DVS1"
 
 	dtask, err := folders.NetworkFolder.CreateDVS(ctx, spec)
 	if err != nil {
@@ -82,16 +82,21 @@ func TestDVS(t *testing.T) {
 		{types.ConfigSpecOperationRemove, "", nil},                            // Remove == OK
 		{types.ConfigSpecOperationAdd, "", nil},                               // Add == OK
 		{types.ConfigSpecOperationAdd, "DVPG0", nil},                          // Add PG == OK
-		{types.ConfigSpecOperationRemove, "", &types.ResourceInUse{}},         // Remove == fail (ResourceInUse)
+		{types.ConfigSpecOperationRemove, "", &types.ResourceInUse{}},         // Remove dvs0 == fail (ResourceInUse)
+		{types.ConfigSpecOperationRemove, "", nil},                            // Remove dvs1 == OK (no VMs attached)
 		{types.ConfigSpecOperationRemove, "", &types.ManagedObjectNotFound{}}, // Remove == fail (ManagedObjectNotFound)
 	}
 
 	for x, test := range tests {
+		dswitch := dvs
+
 		switch test.err.(type) {
 		case *types.ManagedObjectNotFound:
 			for i := range config.Host {
 				config.Host[i].Host.Value = "enoent"
 			}
+		case *types.ResourceInUse:
+			dswitch = dvs0
 		}
 
 		if test.pg == "" {
@@ -99,11 +104,11 @@ func TestDVS(t *testing.T) {
 				config.Host[i].Operation = string(test.op)
 			}
 
-			dtask, err = dvs.Reconfigure(ctx, config)
+			dtask, err = dswitch.Reconfigure(ctx, config)
 		} else {
 			switch test.op {
 			case types.ConfigSpecOperationAdd:
-				dtask, err = dvs.AddPortgroup(ctx, []types.DVPortgroupConfigSpec{{Name: test.pg}})
+				dtask, err = dswitch.AddPortgroup(ctx, []types.DVPortgroupConfigSpec{{Name: test.pg}})
 			}
 		}
 
@@ -115,7 +120,7 @@ func TestDVS(t *testing.T) {
 
 		if test.err == nil {
 			if err != nil {
-				t.Fatal(err)
+				t.Fatalf("%d: %s", x, err)
 			}
 			continue
 		}
@@ -129,12 +134,12 @@ func TestDVS(t *testing.T) {
 		}
 	}
 
-	ports, err := dvs.FetchDVPorts(ctx)
+	ports, err := dvs.FetchDVPorts(ctx, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
-	if len(ports) != 1 {
-		t.Fatalf("expected 1 ports in DVPorts; got %d", len(ports))
+	if len(ports) != 2 {
+		t.Fatalf("expected 2 ports in DVPorts; got %d", len(ports))
 	}
 
 	dtask, err = dvs.Destroy(ctx)

--- a/simulator/finder_test.go
+++ b/simulator/finder_test.go
@@ -67,8 +67,8 @@ func TestFinderVPX(t *testing.T) {
 		{"ManagedObjectList", "/DC0", 1},
 		{"ManagedObjectList", "/F[01]", 2},
 		{"ManagedObjectListChildren", "/*", m.Datacenter + 3},
-		{"ManagedObjectListChildren", "/*/*", 18},
-		{"ManagedObjectListChildren", "/*/*/*", 24},
+		{"ManagedObjectListChildren", "/*/*", 19},
+		{"ManagedObjectListChildren", "/*/*/*", 25},
 		{"FolderList", "/*", m.Folder},
 		{"DatacenterList", "/F0/*", 1},
 		{"DatacenterList", "/DC0", 1},
@@ -91,13 +91,13 @@ func TestFinderVPX(t *testing.T) {
 		{"ClusterComputeResourceList", "/DC0/host/*", m.Cluster},
 		{"HostSystemList", "/DC0/host/*", m.Host + m.ClusterHost},
 		{"HostSystemList", "/F0/DC1/host/F0/*", m.Host + m.ClusterHost},
-		{"HostSystemList", "DC1_H0", 1},             // find . -type HostSystem -name DC1_H0
-		{"ComputeResourceList", "DC1_H0", 1},        // find . -type ComputeResource -name DC1_H0
-		{"ClusterComputeResourceList", "DC1_C0", 1}, // find . -type ClusterComputeResource -name DC1_H0
-		{"NetworkList", "/DC0/network/*", 2 + m.Portgroup},
+		{"HostSystemList", "DC1_H0", 1},                    // find . -type HostSystem -name DC1_H0
+		{"ComputeResourceList", "DC1_H0", 1},               // find . -type ComputeResource -name DC1_H0
+		{"ClusterComputeResourceList", "DC1_C0", 1},        // find . -type ClusterComputeResource -name DC1_H0
+		{"NetworkList", "/DC0/network/*", 3 + m.Portgroup}, // VM Network + DSwitch + DSwitch-Uplinks + m.Portgroup
 		{"NetworkList", "./*", 1},
 		{"NetworkList", "/F0/DC1/network/VM Network", 1},
-		{"NetworkList", "/F0/DC1/network/F0/*", 1 + m.Portgroup},
+		{"NetworkList", "/F0/DC1/network/F0/*", 2 + m.Portgroup},
 		{"NetworkList", "./F0/DC1_DVPG0", 1},
 		{"NetworkList", "./F0/DC1_DVPG0", 1},
 		{"NetworkList", "DC1_DVPG0", 1}, // find . -type Network -name DC1_DVPG0

--- a/simulator/folder.go
+++ b/simulator/folder.go
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"math/rand"
 	"path"
+	"strings"
 
 	"github.com/google/uuid"
 
@@ -513,6 +514,17 @@ func (f *Folder) CreateDVSTask(req *types.CreateDVS_Task) soap.HasFault {
 				ForwardingClass: "etherswitch",
 			}
 		}
+
+		dvs.AddDVPortgroupTask(&types.AddDVPortgroup_Task{
+			Spec: []types.DVPortgroupConfigSpec{{
+				Name: dvs.Name + "-DVUplinks" + strings.TrimPrefix(dvs.Self.Value, "dvs"),
+				DefaultPortConfig: &types.VMwareDVSPortSetting{
+					Vlan: &types.VmwareDistributedVirtualSwitchTrunkVlanSpec{
+						VlanId: []types.NumericRange{{Start: 0, End: 4094}},
+					},
+				},
+			}},
+		})
 
 		return dvs.Reference(), nil
 	})

--- a/simulator/model_test.go
+++ b/simulator/model_test.go
@@ -31,6 +31,10 @@ func compareModel(t *testing.T, m *Model) {
 	pools := (m.Pool * m.Cluster * m.Datacenter) + (m.Host+m.Cluster)*m.Datacenter
 	// root folder + Datacenter folders {host,vm,datastore,network} + top-level folders
 	folders := 1 + (4 * m.Datacenter) + (5 * m.Folder)
+	pgs := m.Portgroup
+	if pgs > 0 {
+		pgs++ // uplinks
+	}
 
 	tests := []struct {
 		expect int
@@ -39,7 +43,7 @@ func compareModel(t *testing.T, m *Model) {
 	}{
 		{m.Datacenter, count.Datacenter, "Datacenter"},
 		{m.Cluster * m.Datacenter, count.Cluster, "Cluster"},
-		{m.Portgroup * m.Datacenter, count.Portgroup, "Portgroup"},
+		{pgs * m.Datacenter, count.Portgroup, "Portgroup"},
 		{m.Datastore * m.Datacenter, count.Datastore, "Datastore"},
 		{hosts, count.Host, "Host"},
 		{vms, count.Machine, "VirtualMachine"},

--- a/simulator/registry.go
+++ b/simulator/registry.go
@@ -40,6 +40,7 @@ var refValueMap = map[string]string{
 	"VirtualMachine":                 "vm",
 	"VirtualMachineSnapshot":         "snapshot",
 	"VmwareDistributedVirtualSwitch": "dvs",
+	"DistributedVirtualSwitch":       "dvs",
 }
 
 // Map is the default Registry instance.

